### PR TITLE
#135 Use max SoC as battery protection limit.

### PR DESCRIPTION
**Pull request by ArdJonker**

### CODE REVIEW

File: constants.py [hunk 1]
```diff
@@ -65,6 +65,10 @@
 
 # Defaults to 0.85
 CHARGER_PLUS_CAR_ROUNDTRIP_EFFICIENCY: float = 0.85
+# Defaults to min current setting of 6A * 240V = 1440W
+CHARGER_MAX_CHARGE_POWER: int = 1440
```
Comment (Flix6x) [inline constants.py:69]: Why so low? I expected 11A to be a more useful default.

Also, I heard some cars "fall asleep" below 6A, so this setting would be taking it right to the edge of what's possible (for some cars).

File: flexmeasures_client.py [hunk 2]
```diff
@@ -275,29 +292,50 @@ def trigger_schedule(self, *args, **fnc_kwargs):
...
                                 f"adjusted to {self.CAR_MIN_SOC_IN_KWH}kWh.")
                         target_soc = self.CAR_MIN_SOC_IN_KWH
 
+                    # The relaxation window is the period before a calendar item where no
+                    # soc_maxima should be sent to allow the schedule to reach a target higher
+                    # than the CAR_MAX_SOC_IN_KWH.
+                    if target_soc > self.CAR_MAX_SOC_IN_KWH:
+                        window_duration = math.ceil((target_soc - self.CAR_MAX_SOC_IN_KWH)/(c.CHARGER_MAX_CHARGE_POWER/1000)*60) + self.WINDOW_SLACK
+                        start_relaxation_window = time_round((target_datetime + timedelta(minutes=-window_duration)), resolution)
+                        self.log(f"Lifting the soc-maxima due to upcomming target, start_relaxation_window: {start_relaxation_window.isoformat()}.")
```
Comment (Flix6x) [inline flexmeasures_client.py:311]: Typo: upcoming

File: flexmeasures_client.py [hunk 3]
```diff
@@ -275,29 +292,50 @@ def trigger_schedule(self, *args, **fnc_kwargs):
...
                         "value": target_soc,
-                        "datetime": target_datetime,
+                        "datetime": target_datetime.isoformat(),
                     }
                 ],
+                "soc-maxima": [
+                    {
+                        "value": self.CAR_MAX_SOC_IN_KWH,
+                        "datetime": dt.isoformat(),
+                    } for dt in [rounded_now + x * resolution for x in range(0, (start_relaxation_window - rounded_now) // resolution)]
```
Comment (Flix6x) [inline flexmeasures_client.py:337]: Nicely coded!

File: flexmeasures_client.py [hunk 4]
```diff
@@ -31,9 +31,18 @@ class FlexMeasuresClient(hass.Hass):
     DELAY_FOR_REATTEMPTS: int  # number of seconds
     CAR_RESERVATION_CALENDAR: str
 
+    ## Battery protection boundries ##
```
Comment (Flix6x) [inline flexmeasures_client.py:34]: Typo: boundaries

File: v2g_globals.py [hunk 5]
```diff
@@ -20,6 +20,26 @@ def initialize(self):
...
+
+        power = int(float(self.args["charger_max_discharging_power"]))
+        # Make sure this value is between 1440 en 22000
+        tmp = max(min(22000, power), 1440)
+        if power != tmp:
+            self.log(f"Max dis-charge power is changed from {power} to {tmp} to stay within boundries.")
+            power = tmp
+        # set constant so that it can be used in calculations
+        c.CHARGER_MAX_DIS_CHARGE_POWER = power
+        self.log(f"v2g_globals max dis-charge power: {c.CHARGER_MAX_DIS_CHARGE_POWER}.")
```
Comment (Flix6x) [inline v2g_globals.py:41]: This seems like a very reasonable policy. To help keep the code quality, I suggest refactoring this to a separate function. The initialisation function is getting too lengthy, and these lines serve a single purpose.

File: constants.py [hunk 6]
```diff
@@ -65,6 +65,10 @@
...
+# Defaults to min current setting of 6A * 240V = 1440W
+CHARGER_MAX_DIS_CHARGE_POWER: int = 1440
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply constants.py:71]: Will change this to 230 V and 1380 Watt

File: constants.py (hunk 1 above)
Comment (ArdJonker) [reply constants.py:69]: This is just to be on the safe side. If a user makes a typo in the secrets file we keep the power low.

File: flexmeasures_client.py [hunk 7]
```diff
@@ -232,7 +244,9 @@ def trigger_schedule(self, *args, **fnc_kwargs):
         # AJO 2023-03-31:
         # ToDo: Would it be more efficient to determine the target every 15/30/60? minutes instead of at every schedule
         # Set default target_soc to 100% one week from now
-        target_datetime = (time_round(datetime.now(tz=pytz.utc), resolution) + timedelta(days=7)).isoformat()
+        target_datetime = (time_round(self.get_now(), resolution) + timedelta(days=7))
+        # By default we assume no calendar item so no relaxation window is needed
+        start_relaxation_window = target_datetime
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply flexmeasures_client.py:249]: It is only overwritten if the target_soc > limit for battery protection (= CAR_MAX_SOC_IN_KWH). Only under this condition the limit needs to be lifted.

File: flexmeasures_client.py [hunk 8]
```diff
@@ -275,29 +292,50 @@ def trigger_schedule(self, *args, **fnc_kwargs):
...
                         self.log(f"Target SoC from calendar too low: {target_soc}, "
                                 f"adjusted to {self.CAR_MIN_SOC_IN_KWH}kWh.")
                         target_soc = self.CAR_MIN_SOC_IN_KWH
 
+                    # The relaxation window is the period before a calendar item where no
+                    # soc_maxima should be sent to allow the schedule to reach a target higher
+                    # than the CAR_MAX_SOC_IN_KWH.
+                    if target_soc > self.CAR_MAX_SOC_IN_KWH:
+                        window_duration = math.ceil((target_soc - self.CAR_MAX_SOC_IN_KWH)/(c.CHARGER_MAX_CHARGE_POWER/1000)*60) + self.WINDOW_SLACK
+                        start_relaxation_window = time_round((target_datetime + timedelta(minutes=-window_duration)), resolution)
```
[the remark this answers is not in the document]
Comment (ArdJonker) [reply flexmeasures_client.py:310]: Will do, i care about your sanity ;-)

File: flexmeasures_client.py (hunk 2 above)
Comment (ArdJonker) [reply flexmeasures_client.py:311]: ✅

File: flexmeasures_client.py (hunk 3 above)
Comment (ArdJonker) [reply flexmeasures_client.py:337]: Simply love it!

File: flexmeasures_client.py (hunk 4 above)
Comment (ArdJonker) [reply flexmeasures_client.py:34]: ✅

File: v2g_globals.py (hunk 5 above)
Comment (ArdJonker) [reply v2g_globals.py:41]: Will do..

**Review by Flix6x — APPROVED** [review]
Thanks for fixing the type's [sic].

**Pull request merged**